### PR TITLE
DDO-2219 Implement tripperware to automatically mask secrets pulled from Vault

### PR DESCRIPTION
**Pull request by choover-broad**
Implement a [tripperware](https://github.com/improbable-eng/go-httpwares) to automatically mask secrets that are retrieved from Vault.

### Why?

Thelma has a number of runtime-dependencies on Vault. (The legacy-migrated `bee seed` command, for example, loads a bunch of service account credentials from Vault and uses them to issue requests to Terra.) Now that we are working on making it possible to run Thelma from GitHub actions on DSP's public repos, we need some protection in place to prevent these Vault secrets from leaking in public GHA logs.

We have two choices -- either (1) migrate all those secrets to GitHub secrets (heavy lift) or (2) add leak protection to Thelma itself. This PR implements option 2.

### Limitations
This tripperware is intended for use with the [KV Secret backend](https://www.vaultproject.io/docs/secrets/kv) we use in DSP. Note that it:
* Only masks secret values, not keys or paths (keys frequently contain common words and we don't want to obfuscate random Thelma log messages that could contain important information)
* Only masks string values. Eg. if it encounters a secret with a map-typed value, it won't attempt to mask it.
* Only masks secret data that is embedded in 2xx responses (secret data that is embedded in Vault error messages will not be masked).
* Does not mask secret data in requests (so if Thelma *writes* a value to Vault, that value will not be masked).

### CODE REVIEW

**Review by em-may — APPROVED** [review]

**Comment by jyang-broad** [comment]
Most SA secrets are saved in a .json format. do we expect this to mask the whole secret as a string?

File: internal/thelma/app/logging/logging.go [hunk 1]
```diff
@@ -40,6 +40,10 @@ type logConfig struct {
 	}
 }
 
+func init() {
+	Bootstrap()
```
Comment (jyang-broad) [inline internal/thelma/app/logging/logging.go:44]: just curious, was this not required before?

File: internal/thelma/clients/vault/masking_round_tripper_test.go [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+			assert.Len(t, secrets, 1)
+			assert.Equal(t, "xyz", secrets[0])
+			maskFnCalled = true
+		},
+	}
+	client.Transport = maskingTransport
+
+	_, err := client.Get(fakeVaultServer.Server().URL + "/v1/secret/foo/bar")
+	require.NoError(t, err)
+	assert.True(t, maskFnCalled)
```
Comment (jyang-broad) [inline internal/thelma/clients/vault/masking_round_tripper_test.go:30]: this test is a little unclear to me. I think it tests
a) a secret exists
b) it is the expected secret "xyz"
c) maskFn has been called.

I'm not sure what to expect for:
a) is the secret still available to thelma
b) is it just the logging that gets masked.
c) what do i expect to see where there used to be i.e. "xyz": "****"

**Review by jyang-broad — APPROVED** [review]
Desc and approach look sane! Would love to see some more tests that confirm expected behavior

File: internal/thelma/clients/vault/masking_round_tripper_test.go (hunk 2 above)
Comment (mflinn-broad) [inline internal/thelma/clients/vault/masking_round_tripper_test.go:30]: Is there anyway to assert the secret is actually being masked in output, not just that the masking function has been called?

**Review by mflinn-broad — APPROVED** [review]
Code lgtm. Is the intention to address some of the limitations in future prs? Specifically the map type secrets, I know there are some of those floating around in vault.

**Comment by choover-broad** [comment]
> Most SA secrets are saved in a .json format. do we expect this to mask the whole secret as a string?

Right, if there's a Vault secret with a key called `sa-key.json` and the value is the entire .json key, the value is what would be redacted. 

If Thelma parsed the JSON and extracted a specific value from it, that value would still be at risk of leaking in logs unless it was explicitly masked. Same goes for values that are stored base64-encoded in Vault and then decoded for use in Thelma.

GitHub Actions secrets [have the same vulnerabilities](https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-secrets), for what its worth.

TL;DR This isn't a magic bullet and we should still be cautious with secret handling but IMO it's better to have it than not.

File: internal/thelma/clients/vault/masking_round_tripper_test.go (hunk 2 above)
Comment (choover-broad) [reply internal/thelma/clients/vault/masking_round_tripper_test.go:30]: So, the actual masking logic lives in the logging package and there are [test cases](https://github.com/broadinstitute/thelma/blob/main/internal/thelma/app/logging/logging_test.go#L142) there that make sure that it works as expected. All the round tripper does is delegate to the `logging.MaskSecret` function.

Comment (choover-broad) [reply internal/thelma/clients/vault/masking_round_tripper_test.go:30]: The logging package is like the only package in Thelma that has global state (passing around logging objects sucks) so reliably integration testing it here could actually be pretty tough.

Comment (choover-broad) [reply internal/thelma/clients/vault/masking_round_tripper_test.go:42]: I added some comments to clarify what the test is doing, see these comments for more context: https://github.com/broadinstitute/thelma/pull/39#discussion_r933533209

File: internal/thelma/app/logging/logging.go (hunk 1 above)
Comment (choover-broad) [reply internal/thelma/app/logging/logging.go:44]: great question. It's called explicitly in an init function that lives in the `app` package: https://github.com/broadinstitute/thelma/blob/main/internal/thelma/app/app.go#L23. This works fine for when we run Thelma as a command, and during bulk integration tests runs (because a subset of the tests do cause the `app` package to be loaded).

But lots of tests, don't depend on the `app` package, but DO depend on logging. This includes tests in the `vault` package.

Adding this to an init function in the logging package allowed me to confirm masking was working in when I manually ran individual unit tests in just the`vault` package:
```
3:11PM DBG automatically masked 1 fields in Vault response
3:11PM DBG GET /v1/secret/foo/bar
3:11PM DBG returning secret foo/bar: map[abc:******]
3:11PM DBG 200 response={"data":{"abc":"******"},"lease_duration":0,"lease_id":"","renewable":false,"request_id":"","warnings":null}
3:11PM DBG automatically masked 1 fields in Vault response
```

Comment (choover-broad) [reply internal/thelma/app/logging/logging.go:44]: I've removed this from the app package's `init` function. It's harmless to run it both places (it's idempotent), but crufty.

**Comment by choover-broad** [comment]
@mflinn-broad 
> Code lgtm. Is the intention to address some of the limitations in future prs or are they not relevant to how thelma is used? Specifically the map type secrets, I know there are some of those floating around in vault.

No plans just because I'm not sure what the best way to handle those is 🤔 

I think the KV store supports arbitrary JSON data structures for field values, not just map[string,string]. So we could try to traverse the data and mask just the leaf strings (ew), or alternately serialize the whole secret into JSON and mask that (ew)...
